jax/_src/lax/parallel.py

```diff
@@ -749,6 +756,70 @@ def _process_axis_index(self, frame):
 batching.BatchTrace.process_axis_index = _process_axis_index  # type: ignore
 
 
+pdot_p = core.Primitive('pdot')
+
+@pdot_p.def_impl
+def _pdot_impl(x, y, *, axis_name, pos_contract, pos_batch):
+  if axis_name: raise NameError(f"unbound axis name: {axis_name[0]}")
+  return lax.dot_general(x, y, [pos_contract, pos_batch])
+
+@pdot_p.def_abstract_eval
+def _pdot_abstract_eval(x, y, *, axis_name, pos_contract, pos_batch):
+  # TODO: avals with names, check inputs are mapped along axis_name, eliminate
+  return lax.dot_general_p.abstract_eval(
+      x, y, dimension_numbers=[pos_contract, pos_batch],
+      precision=None)
+
+def _pdot_vmap_collective_rule(frame, vals_in, dims_in, *, axis_name,
+                               pos_contract, pos_batch):
+  x, y = vals_in
+  x_dim, y_dim = dims_in
+  x_pos_contract, y_pos_contract = pos_contract
+  x_pos_contract = [x_dim] + [d + (d >= x_dim) for d in x_pos_contract]
+  y_pos_contract = [y_dim] + [d + (d >= y_dim) for d in y_pos_contract]
+  x_pos_batch, y_pos_batch = pos_batch
+  x_pos_batch = [d + (d >= x_dim) for d in x_pos_batch]
+  y_pos_batch = [d + (d >= y_dim) for d in y_pos_batch]
+  remaining_axis_names = tuple(n for n in axis_name if n != frame.name)
+  out = pdot_p.bind(x, y, axis_name=remaining_axis_names,
+                    pos_contract=[x_pos_contract, y_pos_contract],
+                    pos_batch=[x_pos_batch, y_pos_batch])
+  return out, None
+batching.collective_rules[pdot_p] = _pdot_vmap_collective_rule
+
+def _pdot_vmap_batching_rule(vals_in, dims_in, *, axis_name, pos_contract,
+                             pos_batch):
+  x, y = vals_in
+  (pos_contract, pos_batch), result_batch_dim = lax._dot_general_batch_dim_nums(
+      (x.ndim, y.ndim), dims_in, [pos_contract, pos_batch])
+  out = pdot_p.bind(x, y, axis_name=axis_name, pos_contract=pos_contract,
+                    pos_batch=pos_batch)
+  return out, result_batch_dim
+batching.primitive_batchers[pdot_p] = _pdot_vmap_batching_rule
+
+def _pdot_translation_rule(c, x, y, *, axis_name, pos_contract, pos_batch,
+                           axis_env, platform):
+  assert axis_name
+  local_out = lax._dot_general_translation_rule(
+      c, x, y, dimension_numbers=[pos_contract, pos_batch], precision=None)
+  out_tup = xla.parallel_translations[psum_p](
+      c, local_out, axis_name=axis_name, axis_index_groups=None,
+      axis_env=axis_env, platform=platform)
+  out, = xla.xla_destructure(c, out_tup)
+  return out
+xla.parallel_translations[pdot_p] = _pdot_translation_rule
+
+def _pdot_transpose_lhs(g, y, *, axis_name, pos_contract, pos_batch):
+  # TODO: avals with names, call pbroadcast with axis_name
```

Comment:
I'm pretty sure that we need to assert that there were no `axis_name`s left. I don't think that our current AD system guarantees that a cotangent of an unmapped aval is also unmapped, so this would be incorrect.

Reply:
Hmm not sure I follow. Can you give an example of the error you're worried about? Also, what do you mean exactly by checking that there were no axis_names left?

Reply:
Since this code is moving fast (with both of us editing it), I'd like to get this change in to avoid rebasing. Moreover it might be easier to think through the AD rules when we can talk with lower latency, i.e. when we can video chat. For those reasons, and because I think I addressed the other review comments, I'm going to consider this merge-able, but I'll be happy to land fixes as we find need for them!